src/java.base/share/classes/sun/launcher/LauncherHelper.java

```diff
@@ -823,42 +822,42 @@ private static Class<?> loadModuleMainClass(String what) {
         }
 
         System.setProperty("jdk.module.main.class", c.getName());
-        return c;
+        builder.mainClass = c;
     }
 
     /**
      * Loads the main class from the class path (LM_CLASS or LM_JAR).
      */
-    private static Class<?> loadMainClass(int mode, String what) {
+    private static void loadMainClass(int mode, String what, MainEntry.Builder builder) {
         // get the class name
         String cn = null;
+        Class<?> mainClass = null;
         // In LM_JAR mode, keep the underlying file open to retain it in
         // the JarFile/ZipFile cache. This will avoid needing to re-parse
         // the central directory when the file is opened on the class path,
         // triggered by Class.forName below.
         JarFile jarFile = null;
-        switch (mode) {
-            case LM_CLASS:
-                cn = what;
-                break;
-            case LM_JAR:
-                try {
-                    jarFile = new JarFile(what);
-                    cn = getMainClassFromJar(jarFile);
-                } catch (IOException ioe) {
-                    abort(ioe, "java.launcher.jar.error1", what);
-                }
-                break;
-            default:
-                // should never happen
-                throw new InternalError("" + mode + ": Unknown launch mode");
-        }
-
-        // load the main class
-        cn = cn.replace('/', '.');
-        Class<?> mainClass = null;
-        ClassLoader scl = ClassLoader.getSystemClassLoader();
         try {
+            switch (mode) {
+                case LM_CLASS:
+                    cn = what;
+                    break;
+                case LM_JAR:
+                    try {
+                        jarFile = new JarFile(what);
+                        cn = getMainClassFromJar(jarFile);
```

Comment:
Why is this switch block moved into the `catch (LinkageError le)` block? Is it for this handling?